### ISSUE TITLE
Create Entity API

### ISSUE THREAD

**Issue by kyao** [usc-isi-i2/datamart-api#144]
GET /entities  -- returns all entity qnodes satisfying the query
Parameters:
Name: name of the entity. Example name=Q115
Label: exact match with the entity label. Example: label=Ethiopia

Example: GET/entities?name=Q115
Returns TSV:
node1   label   node2   id
Q115    label   "Ethiopia"      Q115-label-1

POST /entities -- creates a new entity node
Returns 201 if all the entities are successfully created. If an entity is already defined, it is updated with the posted values.

Example: POST /entities 
With TSV data:
node1   label   node2   id
Q41550  label   "Organisation for Economic Cooperation and Development"
QUNIT-QDPPoverty-002    label   "Fraction of Populace"
QUNIT-QDPPoverty-002    P31     Q47574

DELETE is allowed if entity is not used in any datasets

**Comment by zmbq** [comment]
Update entity endpoint to be similar to the property endpoint - TSV in body, PUT in transaction, IDs by kgtk.

**Comment by zmbq** [comment]
There are two differences between this endpoint and the properties endpoint.

1. In the properties endpoint you can GET /properties/P... and get the property. In entities you must GET /entities?name=Q... The PUT and DELETE verbs behave the same.
2. In properties, POSTing over an existing property fails if the operation should result in an update. In entities, existing entities are just overwritten.

I have updated the endpoint to accept GET /entities/Q... . If you pass both an entity in the URL and in a query `name` parameter, the query parameter is ignored. Should we remove the `name` query parameter?

All other additions from the properties endpoint where added - PUT and POST are wrapped in a transaction, TSVs are read from the request body, kgtk is used to add edges.

**Comment by zmbq** [comment]
Another issue is that GET returns all edges of the Qnode. If the Qnode describes a variable, there can be many edges. However, when updating an entity, only `label`, `description`, `alias` and `description` are allowed - all the others are deleted from the database. Perhaps we should only delete the labels we can update through the endpoint, instead of all of them?

**Comment by kyao** [comment]
> I have updated the endpoint to accept GET /entities/Q... . If you pass both an entity in the URL and in a query `name` parameter, the query parameter is ignored. Should we remove the `name` query parameter?

Yes, `name` query parameter should be removed. Following REST convention we should just use GET /entities/Q...

**Comment by kyao** [comment]
> Another issue is that GET returns all edges of the Qnode. If the Qnode describes a variable, there can be many edges. However, when updating an entity, only `label`, `description`, `alias` and `description` are allowed - all the others are deleted from the database. Perhaps we should only delete the labels we can update through the endpoint, instead of all of them?

Maybe we should let the delete fail if there are other edges. A Qnode should at least have a `label`. If we delete the `label` edge, then the dataset query for may fail.

**Comment by zmbq** [comment]
> > Another issue is that GET returns all edges of the Qnode. If the Qnode describes a variable, there can be many edges. However, when updating an entity, only `label`, `description`, `alias` and `description` are allowed - all the others are deleted from the database. Perhaps we should only delete the labels we can update through the endpoint, instead of all of them?
> 
> Maybe we should let the delete fail if there are other edges. A Qnode should at least have a `label`. If we delete the `label` edge, then the dataset query for may fail.

The edges are deleted in PUT and POST requests as well, not just in DELETE requests. Whenever someone updates an entity, all of its edges are deleted, and new ones are written.

**Comment by zmbq** [comment]
In PUT And POST, only delete the edges that this endpoint supports.
In DELETE - fail if the entity has edges that this endpoint does not support.

**Comment by zmbq** [comment]
Updated in development
